### PR TITLE
Fix use-after-free issue with the global 'default_material' (issue 1483)

### LINKED ISSUE #1483: segmentation fault in _default_material argument of create_structure_and_set_materials
When I ran a simulation case involved gradient calculation (`calculate_gradient()`), `default_material` in C++ implementation is never assigned the type of `material_type` and pointed to a valid memory allocation in the workflow of gradient calculation. So the conversion from `void*` to `material_type` is invalid. A simple workaround is to delete the "else if" branch, and It passed test cases. Please check this issue.

### PR DESCRIPTION
**Pull request by kkg4theweb**
Signed-off-by: Krishna Gadepalli <kkg@google.com>

Closes #1483.

### CODE REVIEW

File: src/material_data.hpp [hunk 1]
```diff
@@ -54,6 +54,18 @@ struct transition {
   }
 
   bool operator!=(const transition &other) const { return !(*this == other); }
+
+  // NOTE: We could add a copy constructor but that requires a lot more
+  // code cleanup!
+  void copy_from(transition& from) {
```
Comment (stevengj) [inline src/material_data.hpp:60]: Make the arguments `const`

File: src/material_data.hpp [hunk 2]
```diff
@@ -227,6 +286,35 @@ struct material_data {
...
     material_grid_kinds = U_DEFAULT;
   }
+
+  void copy_from(material_data& from) {
+    which_subclass = from.which_subclass;
+    medium.copy_from(from.medium);
+
+    user_func = from.user_func;
+    // NOTE: the user_data field here opaque/void - so this is the best we can do.
+    user_data = from.user_data;
```
Comment (stevengj) [inline src/material_data.hpp:296]: This may be problematic — we may need to increment a Python reference count.

Comment (stevengj) [reply src/material_data.hpp:296]: Not a regression, however — filed as #1589

File: src/material_data.hpp [hunk 3]
```diff
@@ -227,6 +286,35 @@ struct material_data {
...
+    memcpy(epsilon_dims, from.epsilon_dims, 3 * sizeof(size_t));
+    if (from.epsilon_data) {
+      int N = from.epsilon_dims[0] * from.epsilon_dims[1] * from.epsilon_dims[2];
+      epsilon_data = new double[N];
+      memcpy(epsilon_data, from.epsilon_data, N * sizeof(double));
+    }
+
+    grid_size = from.grid_size;
+    if (from.weights) {
+      int N = from.grid_size.x * from.grid_size.y * from.grid_size.z;
```
Comment (stevengj) [inline src/material_data.hpp:308]: Should this be `size_t`

Comment (stevengj) [reply src/material_data.hpp:308]: Probably this should just call `update_weights` and we should fix it there.

File: src/material_data.hpp [hunk 4]
```diff
@@ -227,6 +286,35 @@ struct material_data {
...
+    medium.copy_from(from.medium);
+
+    user_func = from.user_func;
+    // NOTE: the user_data field here opaque/void - so this is the best we can do.
+    user_data = from.user_data;
+    do_averaging = from.do_averaging;
+
+    memcpy(epsilon_dims, from.epsilon_dims, 3 * sizeof(size_t));
+    if (from.epsilon_data) {
+      int N = from.epsilon_dims[0] * from.epsilon_dims[1] * from.epsilon_dims[2];
```
Comment (stevengj) [inline src/material_data.hpp:301]: should be `size_t`

File: src/material_data.hpp (hunk 3 above)
Comment (stevengj) [reply src/material_data.hpp:308]: Separately, we should probably add a check that the user `grid_size` makes sense (e.g. is not negative).

Comment (kkg4theweb) [reply src/material_data.hpp]: Done. Made it into size_t

**Pull request merged**